File: C/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int main(void)
{
    int digit, num_len, valid;
    long num;

    // Prompt user input
    num = get_long("Numbers: ");

    // Check for 'num' length
    num_len = floor(log10(num)) + 1;

    // Calculate checksum
    long counter = 10;
    int result = 0;
    int side1 = 0;
    int side2 = num % 10;

    for (int i = 0; i < (num_len - 1); i += 2)
    {
        side1 = ((num / counter) % 10) * 2;
        side2 += ((num / (counter * 10)) % 10);

        // check if 'checksum' has 2 digits
        digit = floor(log10(side1)) + 1;
        ;
        if (digit > 1)
        {
            side1 = (side1 % 10) + ((side1 / 10) % 10);
        }

        result += side1;
        counter *= 100;
    }
    // check validity
    valid = (result += side2) % 10;

    // Check for starting digits
    int start_digit = num / pow(10, (num_len - 2));

    if (valid == 0)
    {
        if ((start_digit == 34 || start_digit == 37) && num_len == 15)
        {
            printf("AMEX\n");
        }
        else if (start_digit > 50 && start_digit < 56 && num_len == 16)
        {
            printf("MASTERCARD\n");
        }
        else if ((start_digit / 10) == 4 && (num_len == 13 || num_len == 16))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: Week06_Python/psets6/sentimental-credit/credit.py
from cs50 import get_int


def main():
    card_num = get_int("Number: ")

    if valid_card_number(card_num):
        print_card_name(card_num)
    else:
        print("INVALID")


def valid_card_number(card_num):
    card_len = len(str(card_num))

    if card_len < 13 or card_len > 16:
        return False

    checksum = 0
    for counter, num in enumerate(reversed(str(card_num))):
        if counter % 2 == 0:
            checksum += int(num)
        else:
            for i in str(int(num) * 2):
                checksum += int(i)

    if checksum % 10 == 0:
        return True
    else:
        return False


def print_card_name(card_num):
    card_len = len(str(card_num))
    start_digit = str(card_num)[:2]

    if (start_digit == "34" or start_digit == "37") and card_len == 15:
        print("AMEX")
    elif start_digit >= "51" and start_digit <= "55" and card_len == 16:
        print("MASTERCARD")
    elif start_digit[:1] == "4" and (card_len == 13 or card_len == 16):
        print("VISA")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
